plugins/pastefromgdocs/plugin.js

```diff
@@ -0,0 +1,51 @@
+﻿/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+( function() {
+	CKEDITOR.plugins.add( 'pastefromgdocs', {
+		requires: 'pastetools',
+
+		init: function( editor ) {
+			var pasteToolsPath = CKEDITOR.plugins.getPath( 'pastetools' ),
+				path = this.path;
+
+			editor.pasteTools.register( {
+				filters: [
+					CKEDITOR.getUrl( pasteToolsPath + 'filter/common.js' ),
+					CKEDITOR.getUrl(  path + 'filter/default.js' )
+				],
+
+				canHandle: function( evt ) {
+					var detectGDocsRegex = /id=(\"|\')docs\-internal\-guid\-/;
+
+					return detectGDocsRegex.test( evt.data.dataValue );
+				},
+
+				handle: function( evt, next ) {
+					var data = evt.data,
+						dataTransferHtml = CKEDITOR.plugins.clipboard.isCustomDataTypesSupported ?
+							data.dataTransfer.getData( 'text/html', true ) : null,
+						gDocsHtml = dataTransferHtml || data.dataValue;
+
+					// Do not apply paste filter to data filtered by the GDocs filter (https://dev.ckeditor.com/ticket/13093).
+					data.dontFilter = true;
+					data.dataValue = CKEDITOR.pasteFilters.gdocs( gDocsHtml, editor );
+
+					if ( editor.config.forcePasteAsPlainText === true ) {
+						// If `config.forcePasteAsPlainText` set to true, force plain text even on Word content (#1013).
+						data.type = 'text';
+					} else if ( !CKEDITOR.plugins.clipboard.isCustomCopyCutSupported && editor.config.forcePasteAsPlainText === 'allow-word' ) {
+						// In browsers using pastebin when pasting from Word, evt.data.type is 'auto' (not 'html') so it gets converted
+						// by 'pastetext' plugin to 'text'. We need to restore 'html' type (#1013) and (#1638).
+						data.type = 'html';
+					}
```

Comment:
Yes, it should be a generic handler, but it's a little bit tricky.

At the moment it's not possible to detect the source of the paste, as this information is not passed in the event. Ideally we can introduce the unified handling for forcing paste type and passing info about the source of paste, e.g.:

```javascript
// PfW handler
editor.pasteTools.register( {
	priority: 10,
	handle( { data } ) {
		if ( isWordContent || editor._.nextPasteType ) {
			data.source = 'word';
		}
	}
} );

// Generic handler in pastetext.
editor.pasteTools.register( {
	priority: 999, // Fire at the end.
	handle( { data } ) {
		if ( editor.config.forcePasteAsPlainText === true ) {
			// If `config.forcePasteAsPlainText` set to true, force plain text even on Word content (#1013).
			data.type = 'text';
		} else if ( !CKEDITOR.plugins.clipboard.isCustomCopyCutSupported && editor.config.forcePasteAsPlainText === 'allow-word' && data.source === 'word' ) {
			// In browsers using pastebin when pasting from Word, evt.data.type is 'auto' (not 'html') so it gets converted
			// by 'pastetext' plugin to 'text'. We need to restore 'html' type (#1013) and (#1638).
			data.type = 'html';
		}
	}
} );

// Generic handler in pastetools.
editor.pasteTools.register( {
	priority: Infinity, // Fire at THE VERY END.
	handle( evt ) {
		delete editor._.nextPasteType;
	}
} );
```

The above listing shows two new mechanisms:

* `evt.data.source` – paste event's data can be extended to contain `source` property, which will indicate the source of paste data (Word in the example). It can be used to detect data format in other plugins.
* `editor._.nextPasteType` – this property will be used to force paste type. It can be added by paste commands (like `pastefromword` one) and removed at the end via generic handler in `pastetools`.

However making such handling generic forces us to refactor most of paste processing in our plugins, which is probably out of scope for 4.13.0.

Reply:
https://github.com/ckeditor/ckeditor-dev/issues/3369